### PR TITLE
Add `torch.flip{lr, ud}`

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #38349

TODO:
* [x] Add Tests

### CODE REVIEW

**Comment by kshitij12345** [comment]
@mruberry Please review.

Will add the test post #37826 as I want to use `_rand_shape` from it.

Thanks:).

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -5821,6 +5821,64 @@ def merge_dicts(*dicts):
              [ 0,  1]]])
 """.format(**common_args))
 
+add_docstr(torch.fliplr,
+           r"""
+fliplr(input) -> Tensor
+
+Flip array in the left/right direction.
```
Comment (mruberry) [inline torch/_torch_docs.py:5828]: I know this is what the NumPy documentation says, which is a smart start, but I'd like to clarify the behavior for what happens when the tensor has more than 2 dimensions. In that case is a "row" what you'd expect?

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -5821,6 +5821,64 @@ def merge_dicts(*dicts):
...
+
+Flip the entries in each row in the left/right direction. 
+Columns are preserved, but appear in a different order than before.
+
+Args:
+    {input}
```
Comment (mruberry) [inline torch/_torch_docs.py:5834]: must be at least 2D

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -5821,6 +5821,64 @@ def merge_dicts(*dicts):
...
+
+Example::
+
+    >>> x = torch.arange(8).view(2, 2, 2)
```
Comment (mruberry) [inline torch/_torch_docs.py:5838]: An additional 2D example may be more natural

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -5821,6 +5821,64 @@ def merge_dicts(*dicts):
...
+
+            [[6, 7],
+             [4, 5]]])
+""".format(**common_args))
+
+add_docstr(torch.flipud,
+           r"""
+flipud(input) -> Tensor
+
+Flip array in the up/down direction.
```
Comment (mruberry) [inline torch/_torch_docs.py:5857]: Same doc comments as above (including the description of the input arg).

**Review by mruberry — CHANGES_REQUESTED** [review]
This is great, thanks @kshitij12345! I made some comments to improve the documentation. Also, I'd like to add more testing. We should add `test_fliplr` and `test_flipud` to `TestTorchDeviceType` in test_torch.py. That will let us test them on CPU and CUDA and on multiple device types, too. Further, that test class defines a helper, `_np_compare,` that you can use to implement most of the test for you. It will compare the results of torch.flip{lr, ud} to NumPy's results. Take a look at some of the tests using `_np_compare` already and let me know if you have any questions.

Once the docs are improved and we get the new tests in I think this PR will be good to go!

File: torch/_torch_docs.py (hunk 1 above)
Comment (mruberry) [reply torch/_torch_docs.py:5828]: Oh! And I almost forgot: we should clarify that PyTorch returns a new tensor:

"Flips array in the left/right direction, returning a new tensor."

File: torch/_torch_docs.py (hunk 4 above)
Comment (mruberry) [reply torch/_torch_docs.py:5857]: We should also clarify here that we return a new tensor (same with fliplr).

**Comment by kshitij12345** [comment]
@mruberry 

* Have updated the docs.

* For adding test to `TestTorchDeviceType`, I am actually waiting for #37826 (which I feel is almost  ready) as I am planning to use `_rand_shape` and `_np_compare_func` (which copies the result np array. This is needed as we are flipping the arrays here which leads to negative strides).

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -5821,6 +5821,56 @@ def merge_dicts(*dicts):
...
+add_docstr(torch.fliplr,
+           r"""
+fliplr(input) -> Tensor
+
+Flip array in the left/right direction, returning a new tensor.
+
+Flip the entries in each row in the left/right direction. 
+Columns are preserved, but appear in a different order than before.
+
+Notes: Equivalent to input[:,::-1]. Requires the array to be at least 2-D.
```
Comment (mruberry) [inline torch/_torch_docs.py:5833]: Note:
  ...

Take a look at torch.bincount for an example.

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -5821,6 +5821,56 @@ def merge_dicts(*dicts):
...
+add_docstr(torch.flipud,
+           r"""
+flipud(input) -> Tensor
+
+Flip array in the up/down direction, returning a new tensor.
+
+Flip the entries in each column in the up/down direction.
+Rows are preserved, but appear in a different order than before.
+
+Notes: Equivalent to input[::-1,...]. Requires the array to be at least 1-D.
```
Comment (mruberry) [inline torch/_torch_docs.py:5858]: Same note formatting as above. 

```
Note:
  ...
```

**Comment by mruberry** [comment]
> @mruberry
> 
> * Have updated the docs.
> * For adding test to `TestTorchDeviceType`, I am actually waiting for #37826 (which I feel is almost  ready) as I am planning to use `_rand_shape` and `_np_compare_func` (which copies the result np array. This is needed as we are flipping the arrays here which leads to negative strides).

Sounds smart. Just re-request when you're ready. I pointed out a couple doc nits and @anjali411 has a suggestion for autograd testing, too.

**Comment by kshitij12345** [comment]
@mruberry @anjali411 
* Have updated the docs.
* Have added tests in `test_torch.py`.
* Have added the functions in whitelist for complex autograd in `test_autograd.py`

Do review. Thanks.

File: test/test_torch.py [hunk 7]
```diff
@@ -8079,6 +8079,40 @@ def test_complex_flip(self, device, dtype):
...
                 self._np_compare_func((torch_fn, np_fn), data, device, dtype)
 
+    def _test_fliplr_flipud(self, funcs, min_dim, max_dim, device, dtype):
+        for dim in range(min_dim, max_dim + 1):
+            shape = self._rand_shape(dim, 5, 10)
+            # Randomly scale the input
+            data = (torch.randn(*shape).to(dtype) * random.randint(50, 100)).tolist()
+            # Use _np_compare_func as it copies the output of np_func,
+            # which takes care of negative strides if present.
+            self._np_compare_func(funcs, data, device, dtype)
```
Comment (mruberry) [inline test/test_torch.py:8089]: This is going to cause a merge conflict with https://github.com/pytorch/pytorch/pull/39179 as is. That PR is landing now so you should be able to rebase on it shortly.

File: test/test_torch.py [hunk 8]
```diff
@@ -8079,6 +8079,40 @@ def test_complex_flip(self, device, dtype):
...
+        with self.assertRaisesRegex(RuntimeError, "Input must be >= 2-d."):
+            torch.fliplr(torch.tensor(42))
+
+    @dtypes(torch.float, torch.double, torch.int16, torch.int32, torch.int64, torch.cfloat, torch.cdouble)
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_flipud(self, device, dtype):
+        funcs = (torch.flipud, np.flipud)
+        self._test_fliplr_flipud(funcs, 1, 4, device, dtype)
+
+    @dtypes(torch.float, torch.double, torch.int16, torch.int32, torch.int64, torch.cfloat, torch.cdouble)
```
Comment (mruberry) [inline test/test_torch.py:8111]: It's great how thorough these are, but it's probably OK just to test float32, int64, and c64.

**Review by mruberry — CHANGES_REQUESTED** [review]
Looks very good. Made a couple notes that we can reduce the test surface a little and should avoid a merge conflict by waiting a moment to rebase this PR. This PR also needs an update to:

- Add the function names to the list of interned strings (see https://github.com/pytorch/pytorch/blob/46447045ea450069ab9a9cbbf71e86110013fb0b/aten/src/ATen/core/aten_interned_strings.h#L329)
- Update tensors.rst (https://github.com/pytorch/pytorch/blob/master/docs/source/tensors.rst) and torch.rst (https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst) to register the docs

Then we'll just need to verify the tests pass.

File: test/test_torch.py (hunk 7 above)
Comment (kshitij12345) [reply test/test_torch.py]: Thanks for the heads up. 

Updated the code.

File: test/test_torch.py (hunk 8 above)
Comment (kshitij12345) [reply test/test_torch.py]: Updated.

**Comment by kshitij12345** [comment]
@mruberry 
Have addressed the comments. PTAL:)

**Comment by kshitij12345** [comment]
Flake8 failure looks non-relevant to the PR.

**Comment by mruberry** [comment]
> Flake8 failure looks non-relevant to the PR.

You're correct. I submitted a fix but you have to rebase to see it.

**Comment by kshitij12345** [comment]
Gentle ping:)

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @kshitij12345!

If you'd like to implement another NumPy function that Torch is missing, maybe maximum and minimum or broadcast_to?

**Comment by kshitij12345** [comment]
Thanks. 

Sure. Would love to try `maximum` and `minimum`.

**Pull request closed without merging**